alica_engine/include/engine/blackboard/Blackboard.h

```diff
@@ -1,47 +1,322 @@
+            if (BB_VALUE_TYPE_NAMES[i] == type) {
+                // i + 1 to account for std::monostate in the variant
+                return i + 1;
+            }
+        }
+        return {};
+    }
+
+    static constexpr const char* BB_VALUE_TYPE_NAMES[] = {"bool", "int64", "double", "std::string"};
+    static constexpr std::size_t BB_VALUE_TYPE_NAMES_SIZE = sizeof(BB_VALUE_TYPE_NAMES) / sizeof(const char*);
+
+    template <bool PARSE_ARGS = false>
+    struct makeBBValueForIndex
+    {
+        template <class... Args>
+        static BBValueType make(std::size_t index, Args&&... args)
+        {
+            // make a BBValueType constructed from a value of type that is at `index` of the variant
+            // The value itself is either constructed from args or parsed from args
+            // Note: index cannot be std::monostate
+            // throws if variant construction fails
+            assert(index > 0);
+            // Note: make index sequence equal to the number of types in BBValueType
+            return makeHelper(index, std::make_index_sequence<BB_VALUE_TYPE_NAMES_SIZE + 2>(), std::forward<Args>(args)...);
+        }
+
+    private:
+        template <class T>
+        struct Parser
+        {
+            auto operator()(const std::string& value) const
+            {
+                YAML::Node node = YAML::Load(value);
+                return node.as<T>();
+            }
+        };
+
+        template <class... Args, std::size_t... Is>
+        static BBValueType makeHelper(std::size_t index, std::index_sequence<Is...>, Args&&... args)
+        {
+            BBValueType vals[] = {makeBBValueIfIndex<Is>::make(index, std::forward<Args>(args)...)...};
+            if (!vals[index].index()) {
+                // variant construction failed, throw exception
+                throw BlackboardException("variant construction failed");
+            }
+            return vals[index];
+        }
+
+        template <std::size_t INDEX>
+        struct makeBBValueIfIndex
+        {
+            using TypeAtIndex = std::decay_t<decltype(std::get<INDEX>(std::declval<BBValueType>()))>;
+
+            template <class... Args>
+            static BBValueType make(std::size_t index, Args&&... args)
+            {
+                // if INDEX == index, make a variant with a value of type that is same as the variant's type at index INDEX, intialized (or parsed) with value
+                // else makes a invalid variant, i.e. with value monostate
+                if constexpr (PARSE_ARGS) {
+                    if constexpr (!std::is_same_v<TypeAtIndex, std::monostate> && !std::is_same_v<TypeAtIndex, std::any>) {
+                        return index == INDEX ? BBValueType{TypeAtIndex{Parser<TypeAtIndex>{}(args)...}} : BBValueType{};
+                    }
+                } else {
+// TODO: figure out a way to disable warnings, caused due to std::visit. Note: really tricky since small changes can cause unexpected behaviour, so best to
```

Comment:
Ideally we should not allow narrowing conversions:  i.e. do not allow mapping float to int.  In fact I think the only conversion worth maybe supporting is int -> double